File: src/imuBias.py
#!/usr/bin/env python
import roslib; roslib.load_manifest('gps_nav')
import rospy, math
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu
from math import cos, sin, sqrt, atan2
import numpy as np
from numpy import array, dot
from numpy.linalg import pinv


def print_x(x):
    print(x[0, 0], x[1, 0], np.degrees(x[2, 0]))


def control_update(x, u, dt):
    """ x is [x, a, hdg], u is [vel, omega] """

    v = u[0]
    w = u[1]
    if w == 0:
        # approximate straight line with huge radius
        w = 1.e-30
    r = v/w # radius


    return x + np.array([[-r*sin(x[2]) + r*sin(x[2] + w*dt)],
                         [ r*cos(x[2]) - r*cos(x[2] + w*dt)],
                         [w*dt]])


a1 = 0.001
a2 = 0.001
a3 = 0.001
a4 = 0.001

sigma_r = 0.1
sigma_h = a_error = np.radians(1)
sigma_s = 0.00001

def normalize_angle(x, index):
    if x[index] > np.pi:
        x[index] -= 2*np.pi
    if x[index] < -np.pi:
        x[index] = 2*np.pi


def ekfloc_predict(x, P, u, dt):

    h = x[2]
    v = u[0]
    w = u[1]

    if w == 0:
        # approximate straight line with huge radius
        w = 1.e-30
    r = v/w # radius

    sinh = sin(h)
    sinhwdt = sin(h + w*dt)
    cosh = cos(h)
    coshwdt = cos(h + w*dt)

    G = array(
       [[1, 0.5*(dt**2), -r*cosh + r*coshwdt],
        [0, 1, -r*sinh + r*sinhwdt],
        [0, 0, 1]])

    V = array(
        [[(-sinh + sinhwdt)/w, v*(sin(h)-sinhwdt)/(w**2) + v*coshwdt*dt/w],
         [(cosh - coshwdt)/w, -v*(cosh-coshwdt)/(w**2) + v*sinhwdt*dt/w],
         [0, dt]])


    # covariance of motion noise in control space
    M = array([[a1*v**2 + a2*w**2, 0],
               [0, a3*v**2 + a4*w**2]])



    x = x + array([[-r*sinh + r*sinhwdt],
                   [r*cosh - r*coshwdt],
                   [w*dt]])

    P = dot(G, P).dot(G.T) + dot(V, M).dot(V.T)

    return x, P

def ekfloc(x, P, u, z, dt):

    h = x[2]
    v = u[0]
    w = u[1]

    if w == 0:
        # approximate straight line with huge radius
        w = 1.e-30
    r = v/w # radius

    sinh = sin(h)
    sinhwdt = sin(h + w*dt)
    cosh = cos(h)
    coshwdt = cos(h + w*dt)

    F = array(
       [[1, 0.5*(dt**2), -r*cosh + r*coshwdt],
        [0, 1, -r*sinh + r*sinhwdt],
        [0, 0, 1]])

    V = array(
        [[(-sinh + sinhwdt)/w, v*(sin(h)-sinhwdt)/(w**2) + v*coshwdt*dt/w],
         [(cosh - coshwdt)/w, -v*(cosh-coshwdt)/(w**2) + v*sinhwdt*dt/w],
         [0, dt]])


    # covariance of motion noise in control space
    M = array([[a1*v**2 + a2*w**2, 0],
               [0, a3*v**2 + a4*w**2]])


    x = x + array([[-r*sinh + r*sinhwdt],
                   [r*cosh - r*coshwdt],
                   [w*dt]])

    P = dot(F, P).dot(F.T) + dot(V, M).dot(V.T)

    R = np.diag([sigma_r**2, sigma_h**2, sigma_s**2])

    z_est = x

    H = array(
        [[1, 0, 0],
         [0, 1, 0],
         [0, 0, 1]])

    S = dot(H, P).dot(H.T) + R

    #print('S', S)
    K = dot(P, H.T).dot(pinv(S))
    y = z - z_est
    normalize_angle(y, 2)
    y = array([y]).T
    #print('y', y)
    
    # Final Estimate
    x = x + dot(K, y)
    I = np.eye(P.shape[0])
    I_KH = I - dot(K, H)
    #print('i', I_KH)

    P = dot(I_KH, P).dot(I_KH.T) + dot(K, R).dot(K.T)

    return x, P


class imuBias(object):
    def __init__(self, t_start, use_ekf_odom=False):
        self.t_start = t_start
        self.acq_time = 0
        self.use_ekf_odom = use_ekf_odom

        self.x = 0
        self.y = 0
        self.theta = 0
        self.a = 0
        self.dt = 0
        self.v, self.w = 0., 0.
        self.odom_ = Odometry()

        self.pub = rospy.Publisher('bias_corrected_odom', Odometry, queue_size=1)
    
    def quaternion_to_euler(self, x, y, z, w):
        ysqr = y * y

        t3 = +2.0 * (w * z + x * y)
        t4 = +1.0 - 2.0 * (ysqr + z * z)
        Z = math.atan2(t3, t4)

        return Z
    
    def euler_to_quaternion(self, theta):
        w = math.cos(theta/2.0)
        x = 0.
        y = 0.
        z = math.sin(theta/2.0)

        return x, y, z, w
    
    def normalize_angle(self, x):
        if x > np.pi:
            x -= 2*np.pi
        if x < -np.pi:
            x = 2*np.pi

    def get_params(self):
        '''
        If no odometer available, use local EKF odometery if specified
        '''
        try:
            rospy.Subscriber('odom', Odometry, self.callback_odom)
            msg = rospy.wait_for_message('odom', Odometry, 1)
        except rospy.exceptions.ROSException:
            self.v = 0.
            self.w = 0.
            if self.use_ekf_odom:
                rospy.Subscriber('odometry/filtered', Odometry, self.callback_odom)
            else:
                self.x = 0
        rospy.Subscriber('imu/data', Imu, self.callback_imu)

    def callback_imu(self, data):
        self.theta = self.quaternion_to_euler(data.orientation.x,
                                         data.orientation.y,
                                         data.orientation.z,
                                         data.orientation.w)
                                         
        self.a = data.linear_acceleration.x
        self.acq_time = data.header.stamp.secs * 1e9 + data.header.stamp.nsecs
        self.dt = self.acq_time - self.t_start
        # print(self.acq_time)
    
    def callback_odom(self, data):
        # These are the measured x and y coorinates of the rover
        self.x = data.pose.pose.position.x
        self.y = data.pose.pose.position.y

        # These are linear and angular input velocities
        self.v = data.twist.twist.linear.x
        self.w = data.twist.twist.angular.z

    def pub_posteriori(self):
        self.get_params()
        self.X = np.array([[self.x, self.a, self.theta]]).T
        self.U = np.array([self.v, self.w])
        self.P = np.diag([1., 1., 1.])

        xp = self.X.copy()
        xp, _ = ekfloc_predict(xp, self.P, self.U, self.dt) # Prediction
        self.X, self.P = ekfloc(self.X, self.P, self.U, xp, self.dt) # Update
        self.t_start = self.acq_time

        self.odom_.header.stamp.secs = int(self.acq_time * 1e-9)
        self.odom_.header.stamp.nsecs = self.acq_time - self.odom_.header.stamp.secs * 1e9
        self.odom_.pose.pose.position.x = self.x
        self.odom_.pose.pose.position.y = self.y

        # print("dt: ", self.dt)
        # print("Estimated theta: ", self.X[2,1])
        x,y,z,w = self.euler_to_quaternion(self.X[2,1])
        self.odom_.pose.pose.orientation.x = x
        self.odom_.pose.pose.orientation.y = y
        self.odom_.pose.pose.orientation.z = z
        self.odom_.pose.pose.orientation.w = w

        self.odom_.twist.twist.linear.x = self.v
        self.odom_.twist.twist.angular.z = self.w

        self.pub.publish(self.odom_)

if __name__ == '__main__':
    rospy.init_node('imu_bias', anonymous=False)
    rate = rospy.Rate(30)

    t_now = rospy.get_rostime().secs * 1e9 + rospy.get_rostime().nsecs
    imu_bias_obj = imuBias(t_now)

    while not rospy.is_shutdown():
        imu_bias_obj.pub_posteriori()
        rate.sleep()
        # print(rospy.get_rostime().secs * 1e9 + rospy.get_rostime().nsecs)

File: src/imuBias.cpp
#include <ros/ros.h>
#include <nav_msgs/Odometry.h>
#include <sensor_msgs/Imu.h>
#include <Eigen/Dense>
#include <cmath>
#include <tf/tf.h>
#include <iostream>

using namespace Eigen;

// Global variables
double a1 = 0.001;
double a2 = 0.001;
double a3 = 0.001;
double a4 = 0.001;

double sigma_r = 0.1;
double sigma_h = M_PI / 180.0; // 1 degree in radians
double sigma_s = 0.00001;

// Normalize an angle in the state vector x, at index `index`
void normalize_angle(Vector3d& x, int index) {
    if (x(index) > M_PI) {
        x(index) -= 2 * M_PI;
    }
    if (x(index) < -M_PI) {
        x(index) += 2 * M_PI;
    }
}

Vector3d control_update(const Vector3d& x, const Vector2d& u, double dt) {
    double v = u(0);
    double w = u(1);
    if (w == 0) {
        w = 1.e-30;
    }
    double r = v / w;

    Vector3d x_new;
    x_new << -r * sin(x(2)) + r * sin(x(2) + w * dt),
             r * cos(x(2)) - r * cos(x(2) + w * dt),
             w * dt;

    return x + x_new;
}

void ekfloc_predict(Vector3d& x, Matrix3d& P, const Vector2d& u, double dt) {
    double h = x(2);
    double v = u(0);
    double w = u(1);

    if (w == 0) {
        w = 1.e-30;
    }
    double r = v / w;

    double sinh = sin(h);
    double sinhwdt = sin(h + w * dt);
    double cosh = cos(h);
    double coshwdt = cos(h + w * dt);

    Matrix3d G;
    G << 1, 0.5 * (dt * dt), -r * cosh + r * coshwdt,
         0, 1, -r * sinh + r * sinhwdt,
         0, 0, 1;

    Matrix<double, 3, 2> V;
    V << (-sinh + sinhwdt) / w, v * (sin(h) - sinhwdt) / (w * w) + v * coshwdt * dt / w,
         (cosh - coshwdt) / w, -v * (cosh - coshwdt) / (w * w) + v * sinhwdt * dt / w,
         0, dt;

    Matrix2d M;
    M << a1 * v * v + a2 * w * w, 0,
         0, a3 * v * v + a4 * w * w;

    x += Vector3d(
        -r * sinh + r * sinhwdt,
         r * cosh - r * coshwdt,
         w * dt
    );
    P = G * P * G.transpose() + V * M * V.transpose();
}

void ekfloc(Vector3d& x, Matrix3d& P, const Vector2d& u, const Vector3d& z, double dt) {
    double h = x(2);
    double v = u(0);
    double w = u(1);

    if (w == 0) {
        w = 1.e-30;
    }
    double r = v / w;

    double sinh = sin(h);
    double sinhwdt = sin(h + w * dt);
    double cosh = cos(h);
    double coshwdt = cos(h + w * dt);

    Matrix3d F;
    F << 1, 0.5 * (dt * dt), -r * cosh + r * coshwdt,
         0, 1, -r * sinh + r * sinhwdt,
         0, 0, 1;

    Matrix<double, 3, 2> V;
    V << (-sinh + sinhwdt) / w, v * (sin(h) - sinhwdt) / (w * w) + v * coshwdt * dt / w,
         (cosh - coshwdt) / w, -v * (cosh - coshwdt) / (w * w) + v * sinhwdt * dt / w,
         0, dt;

    Matrix2d M;
    M << a1 * v * v + a2 * w * w, 0,
         0, a3 * v * v + a4 * w * w;

    x += Vector3d(
        -r * sinh + r * sinhwdt,
         r * cosh - r * coshwdt,
         w * dt
    );
    P = F * P * F.transpose() + V * M * V.transpose();

    Matrix3d R = Matrix3d::Zero();
    R.diagonal() << sigma_r * sigma_r, sigma_h * sigma_h, sigma_s * sigma_s;

    Vector3d z_est = x;
    Matrix3d H = Matrix3d::Identity();
    Matrix3d S = H * P * H.transpose() + R;
    Matrix3d K = P * H.transpose() * S.inverse();

    Vector3d y = z - z_est;
    normalize_angle(y, 2);
    x += K * y;

    Matrix3d I = Matrix3d::Identity();
    Matrix3d I_KH = I - K * H;
    P = I_KH * P * I_KH.transpose() + K * R * K.transpose();
}

class ImuBias {
public:
    ImuBias(double t_start, bool use_ekf_odom = false)
        : t_start_(t_start), use_ekf_odom_(use_ekf_odom), prev_acq_time_(0.0)
    {
        // Subscribe once
        odom_sub_ = nh_.subscribe("odom", 1, &ImuBias::callback_odom, this);
        imu_sub_  = nh_.subscribe("imu/data", 1, &ImuBias::callback_imu, this);

        // (Optional) Wait for first Odom message
        try {
            ros::topic::waitForMessage<nav_msgs::Odometry>("odom", nh_, ros::Duration(1));
        } catch (const ros::Exception& e) {
            ROS_WARN("No odom message received within 1 second.");
            v_ = 0.0;
            w_ = 0.0;
            if (use_ekf_odom_) {
                odom_sub_ = nh_.subscribe("odometry/filtered", 1, &ImuBias::callback_odom, this);
            } else {
                x_ = 0.0;
            }
        }

        // Publish corrected odometry
        pub_ = nh_.advertise<nav_msgs::Odometry>("bias_corrected_odom", 1);
        // Publish predicted IMU
        pub_imu_ = nh_.advertise<sensor_msgs::Imu>("bias_corrected_imu", 1);
    }

    // IMU callback
    void callback_imu(const sensor_msgs::Imu::ConstPtr& data)
    {
        theta_ = tf::getYaw(data->orientation);
        a_ = data->linear_acceleration.y;

        double current_time = data->header.stamp.toSec();
        if (prev_acq_time_ == 0.0) {
            dt_ = 0.0; // First measurement
        } else {
            dt_ = current_time - prev_acq_time_;
        }
        acq_time_      = current_time;
        prev_acq_time_ = current_time;
    }

    // Odom callback
    void callback_odom(const nav_msgs::Odometry::ConstPtr& data)
    {
        x_ = data->pose.pose.position.x;
        y_ = data->pose.pose.position.y;
        v_ = data->twist.twist.linear.x;
        w_ = data->twist.twist.angular.z;
    }

    // Main update
    void pub_posteriori()
    {
        // Prepare state & control vectors
        X_ << x_, a_, theta_;
        U_ << v_, w_;
        P_ = Matrix3d::Identity();

        // Predict step
        Vector3d xp = X_;
        ekfloc_predict(xp, P_, U_, dt_);

        // Update step
        z_ << x_, a_, theta_;
        ekfloc(X_, P_, U_, z_, dt_);

        t_start_ = acq_time_;  // store last IMU time

        // Publish corrected odom
        publishOdom();

        // Publish predicted IMU
        publishImu();
    }

private:
    void publishOdom()
    {
        odom_.header.stamp = ros::Time::now();
        odom_.pose.pose.position.x = x_;
        odom_.pose.pose.position.y = y_;

        tf::Quaternion q = tf::createQuaternionFromYaw(X_(2));
        odom_.pose.pose.orientation.x = q.x();
        odom_.pose.pose.orientation.y = q.y();
        odom_.pose.pose.orientation.z = q.z();
        odom_.pose.pose.orientation.w = q.w();

        odom_.twist.twist.linear.x = v_;
        odom_.twist.twist.angular.z = w_;

        pub_.publish(odom_);
    }

    void publishImu()
    {
        // Construct an Imu message with EKF-predicted orientation & acceleration
        sensor_msgs::Imu imu_msg;
        imu_msg.header.stamp = ros::Time::now();

        // Orientation from X_(2) = yaw
        tf::Quaternion q = tf::createQuaternionFromYaw(X_(2));
        imu_msg.orientation.x = q.x();
        imu_msg.orientation.y = q.y();
        imu_msg.orientation.z = q.z();
        imu_msg.orientation.w = q.w();

        // We only have linear accel in X_(1) = a_
        // For a simple example, assume all accel in X direction
        imu_msg.linear_acceleration.x = 0.0;
        imu_msg.linear_acceleration.y = a_;
        imu_msg.linear_acceleration.z = 0.0;

        // (Optional) Covariances
        // imu_msg.orientation_covariance[0] = 1e-3;
        // imu_msg.angular_velocity_covariance[0] = -1.0; // no estimate
        // imu_msg.linear_acceleration_covariance[0] = 1e-3;

        pub_imu_.publish(imu_msg);
    }

    ros::NodeHandle nh_;
    // Publishers
    ros::Publisher pub_;      // bias_corrected_odom
    ros::Publisher pub_imu_;  // bias_corrected_imu

    // Subscribers
    ros::Subscriber odom_sub_;
    ros::Subscriber imu_sub_;

    double t_start_;
    bool   use_ekf_odom_;

    // Timing
    double acq_time_ = 0.0;
    double prev_acq_time_ = 0.0;
    double dt_ = 0.0;

    // EKF state
    double x_ = 0.0;
    double y_ = 0.0;
    double theta_ = 0.0;
    double a_ = 0.0;
    double v_ = 0.0;
    double w_ = 0.0;

    nav_msgs::Odometry odom_;

    Vector3d X_;
    Vector3d z_;
    Vector2d U_;
    Matrix3d P_;
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "imu_bias");
    ros::NodeHandle nh;

    double t_now = ros::Time::now().toSec() * 1e9 + ros::Time::now().toNSec();
    ImuBias imu_bias_obj(t_now);

    ros::Rate rate(30);
    while (ros::ok()) {
        // Process callbacks (IMU + Odom)
        ros::spinOnce();

        // Perform the EKF update & publish
        imu_bias_obj.pub_posteriori();

        rate.sleep();
    }
    return 0;
}
